README.md
# airpyllution
[![codecov](https://codecov.io/gh/UBC-MDS/airpyllution/branch/main/graph/badge.svg?token=c6vEGpbs3h)](https://codecov.io/gh/UBC-MDS/airpyllution)
[![build](https://github.com/UBC-MDS/airpyllution/actions/workflows/build.yml/badge.svg)](https://github.com/UBC-MDS/airpyllution/actions/workflows/build.yml)
[![deploy](https://github.com/UBC-MDS/airpyllution/actions/workflows/deploy.yml/badge.svg)](https://github.com/UBC-MDS/airpyllution/actions/workflows/deploy.yml)

`airpyllution` is a Python package for visualizing or obtaining future, historic and current air pollution data using the [OpenWeather API](https://openweathermap.org). Our goal is to enable users the ability to explore air pollution levels in locations around the world by providing visual charts and graphs. We make the data accessible and easy to comprehend in just a few lines of code.

Although there is an abundance of python weather packages and APIs in the Python ecosystem (e.g. [python-weather](https://pypi.org/project/python-weather/), [weather-forecast](https://pypi.org/project/weather-forecast/)), this particular package looks at specifically air pollution data and uses the [Air Pollution API](https://openweathermap.org/api/air-pollution) from OpenWeather. This is a unique package which provides simple and easy to use functions and allows users to quickly access and visualise data.

The data returned from the API includes the polluting gases such as Carbon monoxide (CO), Nitrogen monoxide (NO), Nitrogen dioxide (NO2), Ozone (O3), Sulphur dioxide (SO2), Ammonia (NH3), and particulates (PM2.5 and PM10).

Using the OpenWeatherMap API requires sign up to gain access to an API key.   
For more information about API call limits and API care recommendations please visit the [OpenWeather how to start](https://openweathermap.org/appid) page.
## Functions
This package contains 3 functions: 
- `get_air_pollution()`
- `get_pollution_history()`
- `get_pollution_forecast()`

### `get_air_pollution()`
Fetches the air pollution levels based on a location. Based on the values of the polluting gases, this package uses the [Air Quality Index](https://en.wikipedia.org/wiki/Air_quality_index#CAQI) to determine the level of pollution for the location and produces a coloured map of the area displaying the varying regions of air quality.

### `get_pollution_history()`
Requires a start and end date and fetches historic air pollution data for a specific location. The function returns a data frame with the values of the polluting gases over the specified date range.

### `get_pollution_forecast()`
Fetches air pollution data for the next 5 days for a specific location. The function returns a time series plot of the predicted pollution levels.

## Installation

```bash
$ pip install airpyllution
```
## Usage and Example
[![readthedocs](https://readthedocs.org/projects/pip/badge/?version=latest)](https://airpyllution.readthedocs.io/en/latest/)

1. Create an [OpenWeather API Key](https://openweathermap.org/appid)
2. Install airpyllution
3. Refer to [ReadTheDocs](https://airpyllution.readthedocs.io/en/latest/) for a usage guide and examples.

To use the package, import the package with the following commands:
```
from airpyllution.airpyllution import get_air_pollution
from airpyllution.airpyllution import get_pollution_history
from airpyllution.airpyllution import get_pollution_forecast
```

**Retrieve historic pollution data with specified date range and location:**
```
get_pollution_history(1606488670, 1606747870, 49.28, 123.12, api_key)
```

**Generate an interactive map containing current pollution data by location:**

```
get_air_pollution(49.28, 123.12, api_key, "Current Air Pollution")
```

![](docs/air-pollution-map.png)

**Generate a time-series line chart of forecasted air pollution data:**
```
import altair as alt
alt.renderers.enable("html");

get_pollution_forecast(49.28, 123.12, api_key)
```
![](docs/forecast-example.png)

## Contributors 
- Christopher Alexander (@christopheralex)
- Daniel King (@danfke)
- Mel Liow (@mel-liow)
## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.
## License

`airpyllution` was created by Christopher Alexander, Daniel King, Mel Liow. It is licensed under the terms of the [Hippocratic License 3.0](https://github.com/UBC-MDS/airpyllution/blob/main/LICENSE).

## Credits

`airpyllution` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v2.0.1 (2022-02-05)
### Documentation
* Updates readme for License link ([`394fd1d`](https://github.com/UBC-MDS/airpyllution/commit/394fd1d27839b060520dcfdde7d08a1b7320c0e9))
* Updates license to hippocratic license. ([`fd97073`](https://github.com/UBC-MDS/airpyllution/commit/fd970730006740397242857d86babbaafac704cc))

## v2.0.0 (2022-01-29)
### Feature
* Updates README for package documentation. ([`08b4e9e`](https://github.com/UBC-MDS/airpyllution/commit/08b4e9e25e322586a8f5bb5199566e060e22ad2d))

### Fix
* Typo ([`3c72bd1`](https://github.com/UBC-MDS/airpyllution/commit/3c72bd14487742adc0f374c54ed40d85e391dfb8))

### Breaking
* Format of forecast function has changed.  ([`08b4e9e`](https://github.com/UBC-MDS/airpyllution/commit/08b4e9e25e322586a8f5bb5199566e060e22ad2d))

### Documentation
* Actually make it bold ([`d4c8ef8`](https://github.com/UBC-MDS/airpyllution/commit/d4c8ef88e0273d06cf9eb1b189f20c9e3ec93159))
* Make text bold ([`634c780`](https://github.com/UBC-MDS/airpyllution/commit/634c78057c8f64491d5a51536a00cec9d6b4764a))
* Update readme ([`b8d1557`](https://github.com/UBC-MDS/airpyllution/commit/b8d15577b7f59a82175e86ecf734c738fdf1983a))

## v1.1.3 (2022-01-29)
### Documentation
* Hide output of key cell ([`868fa37`](https://github.com/UBC-MDS/airpyllution/commit/868fa374c5daa6bbc396751786a3e9e76c8ddb03))

## v1.1.2 (2022-01-29)
### Documentation
* Add hide-input to api key cell ([`dbe5c20`](https://github.com/UBC-MDS/airpyllution/commit/dbe5c204981445e4c7b248d0b9cf9a262a33d9fb))

## v1.1.1 (2022-01-29)
### Documentation
* Remove API key and put import it from an external file ([`6d2efde`](https://github.com/UBC-MDS/airpyllution/commit/6d2efded311c25e6bfadda4118c1777ca175eca0))

## v1.1.0 (2022-01-29)
### Feature
* Updates release.yml ([`0c41bfd`](https://github.com/UBC-MDS/airpyllution/commit/0c41bfd8884bbb813b057012dd37d00a0dab156f))
* Adds CD pipeline ([`7021a8c`](https://github.com/UBC-MDS/airpyllution/commit/7021a8c4d91efe84823b07c959fa9dbeede7a7fb))
* Add Exceptions to function ([`efcbdc8`](https://github.com/UBC-MDS/airpyllution/commit/efcbdc84502d714efa755335cc65c90a9fc63549))
* Fix tests ([`e6e2d1c`](https://github.com/UBC-MDS/airpyllution/commit/e6e2d1cbfe1581a7ae95959490dec019b60bf363))
* Update ci-cd.yml ([`f782dda`](https://github.com/UBC-MDS/airpyllution/commit/f782dda23359d7803d2d63ec8854e1c3da5c538b))
* Add imports, remove print statement ([`7ea2f4d`](https://github.com/UBC-MDS/airpyllution/commit/7ea2f4d69ae0d5b336933628918df9fc3608c69c))
* Add test for api error ([`538d241`](https://github.com/UBC-MDS/airpyllution/commit/538d2418494032c8260e579a4f781fb811ad60f6))
* Refactor try-except ([`5933e1e`](https://github.com/UBC-MDS/airpyllution/commit/5933e1e4a11ee95ccb5e9cbf30096b2d89095004))
* Add docstring to mocked_requests_get_pollution function ([`70fcf7e`](https://github.com/UBC-MDS/airpyllution/commit/70fcf7ea571bdac1aeaad7e83fd25cdcbdb80aee))
* Fixes package issues for testPypI ([`a06b378`](https://github.com/UBC-MDS/airpyllution/commit/a06b378e902e09573654d66f0163236d401efdd6))
* Correct return message of isinstance in get_air_pollution ([`0b3c117`](https://github.com/UBC-MDS/airpyllution/commit/0b3c1172ec1f63acfba70cea06508d96cf4eb2a2))
* Modify data conversion function to work with all functions that call it ([`1cf707f`](https://github.com/UBC-MDS/airpyllution/commit/1cf707f52923a8bbd506e4b2e9b68c8022f785c5))
* Modify url variables to work with all functions ([`13a992e`](https://github.com/UBC-MDS/airpyllution/commit/13a992eeb811d6653c1af76a6fe44f51e70a05ed))
* Create fully functional get_air_pollution function ([`e9e23a9`](https://github.com/UBC-MDS/airpyllution/commit/e9e23a9ab406e5c661115fc7f8dd4092aa7cfed5))
* Updates forecast function to use utils.py ([`55b2915`](https://github.com/UBC-MDS/airpyllution/commit/55b291570bf6236a77a9a3442174b2fd9c1c66b5))
* Add draft of get_air_pollution ([`b167a08`](https://github.com/UBC-MDS/airpyllution/commit/b167a0811556f61d1e011144aec2949dabcd9b90))
* Fix global url ([`c6b8d10`](https://github.com/UBC-MDS/airpyllution/commit/c6b8d1037afcd9585c84e69595a291cbc6cdb8be))
* Move mock data to constants ([`565b3e6`](https://github.com/UBC-MDS/airpyllution/commit/565b3e62fa66d56254e75db9cb825064ab5ce836))
* Refactor tests ([`2a7c9bc`](https://github.com/UBC-MDS/airpyllution/commit/2a7c9bc128e498bc880362afe2a9831c0fbcd747))
* Add mock data to constants ([`d8938bd`](https://github.com/UBC-MDS/airpyllution/commit/d8938bd15a54443becbaa62ebfab91f23a402043))
* Change request method to GET ([`8bac5bf`](https://github.com/UBC-MDS/airpyllution/commit/8bac5bfa220d8db52be5ca771e16ebdba03de3f2))
* Add convert data to pandas function to utils ([`81395c4`](https://github.com/UBC-MDS/airpyllution/commit/81395c46f59f8573a8e2aa1b6f5e3a902885beea))
* Adds defensive tests for lat and lon in forecast ([`d27a727`](https://github.com/UBC-MDS/airpyllution/commit/d27a727abc6a5c1083905019ae6cadc1914d571e))
* Add back api_key ([`5a12c6f`](https://github.com/UBC-MDS/airpyllution/commit/5a12c6f14cce5f19f76015d482d039243fa36ed6))
* Add exception handling to return api error response to client ([`9424bef`](https://github.com/UBC-MDS/airpyllution/commit/9424bef3073794542f11f7128fc8fff14bfd5b3b))
* Clean up and add utils module ([`5c96b5b`](https://github.com/UBC-MDS/airpyllution/commit/5c96b5b8fd257d9de0699488ec6382a197f4db72))
* Add unit tests and exception handling ([`61f30c2`](https://github.com/UBC-MDS/airpyllution/commit/61f30c2c47937019229983c106b9b96979c8ad02))
* Add a working unit test ([`4af9970`](https://github.com/UBC-MDS/airpyllution/commit/4af997099add9850c0b6afe403021658e7ec3378))
* Working test ([`ca816a8`](https://github.com/UBC-MDS/airpyllution/commit/ca816a8e9e6ef3b1ce2d46b66b9fb95826446b9b))
* Add wip test ([`3a36a77`](https://github.com/UBC-MDS/airpyllution/commit/3a36a77356807af29167986cf130ec68f0a56473))

tests/constants.py
mock_history_data = {
    "coord": [50.0, 50.0],
    "list": [
        {
            "main": {"aqi": 2},
            "components": {
                "co": 270.367,
                "no": 5.867,
                "no2": 43.184,
                "o3": 4.783,
                "so2": 14.544,
                "pm2_5": 13.448,
                "pm10": 15.524,
                "nh3": 0.289,
            },
            "dt": 1606482000,
        },
        {
            "main": {"aqi": 2},
            "components": {
                "co": 280.38,
                "no": 8.605,
                "no2": 42.155,
                "o3": 2.459,
                "so2": 14.901,
                "pm2_5": 15.103,
                "pm10": 17.249,
                "nh3": 0.162,
            },
            "dt": 1606478400,
        },
    ],
}

mock_pollution_data = {
    "coord": {"lon": 123.12, "lat": 49.28},
    "list": [
        {
            "main": {"aqi": 1},
            "components": {
                "co": 310.42,
                "no": 0,
                "no2": 11.14,
                "o3": 35.76,
                "so2": 0.58,
                "pm2_5": 3.33,
                "pm10": 4.76,
                "nh3": 0.1,
            },
            "dt": 1642748400,
        }
    ],
}

mock_forecast_data = {
    "coord": [50.0, 50.0],
    "list": [
        {
            "main": {"aqi": 2},
            "components": {
                "co": 270.367,
                "no": 5.867,
                "no2": 43.184,
                "o3": 4.783,
                "so2": 14.544,
                "pm2_5": 13.448,
                "pm10": 15.524,
                "nh3": 0.289,
            },
            "dt": 1606482000,
        },
        {
            "main": {"aqi": 2},
            "components": {
                "co": 280.38,
                "no": 8.605,
                "no2": 42.155,
                "o3": 2.459,
                "so2": 14.901,
                "pm2_5": 15.103,
                "pm10": 17.249,
                "nh3": 0.162,
            },
            "dt": 1606478400,
        },
    ],
}

mock_api_invalid_key_error = {
    "cod": 401,
    "message": "Invalid API key. Please see \
    http://openweathermap.org/faq#error401 for more info.",
}

mock_params = {
    "lat": 49.28,
    "lon": 123.12,
    "start": 1606488670,
    "end": 1606747870,
    "appid": "mock_api_key",
}

mock_incorrect_params = {
    "lat": "latitude_val",
    "lon": "longitude_val",
    "lat_oor": -100.0,
    "lon_oor": 181.0,
    "start": 1234.567,
    "end": 3.14159,
    "appid": "invalid_api_key",
}

mock_error_params = {
    "lat": "latitude_val",
    "lon": "longitude_val",
    "lat_oor": -100.0,
    "lon_oor": 181.0,
    "start": 1234.567,
    "end": 3.14159,
    "appid": "api_error",
}

mock_invalid_message = "Invalid API key. Please see \
    http://openweathermap.org/faq#error401 for more info."


tests/test_AirPyllution.py
from airpyllution import airpyllution


tests/test_airpyllution.py
from airpyllution import airpyllution
from pandas._testing import assert_frame_equal
from unittest.mock import patch
from constants import (
    mock_api_invalid_key_error,
    mock_error_params,
    mock_forecast_data,
    mock_history_data,
    mock_incorrect_params,
    mock_params,
    mock_pollution_data,
    mock_invalid_message,
)
from airpyllution.utils import convert_data_to_pandas
import altair as alt
from math import floor


def mocked_requests_get_pollution(*args, **kwargs):
    """
    Function for mocking a Response object.
    This intercepts any API calls when called within the test suite and returns
    an instance of a MockResponse object. This should be used with
    the @patch decorator.
    """

    class MockResponse:
        def __init__(self, json_data, status_code):
            self.json_data = json_data
            self.status_code = status_code

        def json(self):
            return self.json_data

    if (
        kwargs["url"]
        == "http://api.openweathermap.org/data/2.5/air_pollution/history"
    ):

        if kwargs["params"]["appid"] == "invalid_api_key":
            return MockResponse(mock_api_invalid_key_error, 404)

        if kwargs["params"]["appid"] == "api_error":
            return "ERROR"

        return MockResponse(mock_history_data, 200)

    elif (
        kwargs["url"] == "http://api.openweathermap.org/data/2.5/air_pollution"
    ):

        if kwargs["params"]["appid"] == "invalid_api_key":
            return MockResponse(mock_api_invalid_key_error, 404)

        if kwargs["params"]["appid"] == "api_error":
            return "ERROR"

        return MockResponse(mock_pollution_data, 200)

    elif (
        kwargs["url"]
        == "http://api.openweathermap.org/data/2.5/air_pollution/forecast"
    ):

        if kwargs["params"]["appid"] == "invalid_api_key":
            return MockResponse(mock_api_invalid_key_error, 404)

        if kwargs["params"]["appid"] == "api_error":
            return "ERROR"
        return MockResponse(mock_forecast_data, 200)

    return MockResponse(
        {"cod": 401, "message": mock_invalid_message},
        404,
    )


@patch("requests.get", side_effect=mocked_requests_get_pollution)
def test_pollution_history(mock_api_call):
    """Test fetching pollution history from API"""

    # Invalid input type
    assert (
        airpyllution.get_pollution_history(
            mock_incorrect_params["start"],
            mock_params["end"],
            mock_params["lat"],
            mock_params["lon"],
            mock_params["appid"],
        )
        == "start_date input should be an int"
    )

    # Invalid input type
    assert (
        airpyllution.get_pollution_history(
            mock_params["start"],
            mock_incorrect_params["end"],
            mock_params["lat"],
            mock_params["lon"],
            mock_params["appid"],
        )
        == "end_date input should be an int"
    )

    # Invalid input type
    assert (
        airpyllution.get_pollution_history(
            mock_params["start"],
            mock_params["end"],
            mock_incorrect_params["lat"],
            mock_params["lon"],
            mock_params["appid"],
        )
        == "Latitude input should be a float"
    )

    # Invalid input type
    assert (
        airpyllution.get_pollution_history(
            mock_params["start"],
            mock_params["end"],
            mock_params["lat"],
            mock_incorrect_params["lon"],
            mock_params["appid"],
        )
        == "Longitude input should be a float"
    )

    # API error
    assert (
        airpyllution.get_pollution_history(
            mock_params["start"],
            mock_params["end"],
            mock_params["lat"],
            mock_params["lon"],
            mock_error_params["appid"],
        )
        == "An error occurred requesting data from the API"
    )

    # Invalid API key, tests nested try-except
    assert (
        airpyllution.get_pollution_history(
            mock_params["start"],
            mock_params["end"],
            mock_params["lat"],
            mock_params["lon"],
            mock_incorrect_params["appid"],
        )
        == mock_invalid_message
    )

    pollution_data_frame = airpyllution.get_pollution_history(
        mock_params["start"],
        mock_params["end"],
        mock_params["lat"],
        mock_params["lon"],
        mock_params["appid"],
    )

    assert_frame_equal(
        pollution_data_frame, convert_data_to_pandas(mock_history_data)
    )


@patch("requests.get", side_effect=mocked_requests_get_pollution)
def test_air_pollution(mock_api_call):
    """Tests get air pollution function"""

    # Invalid input type
    assert (
        airpyllution.get_air_pollution(
            mock_incorrect_params["lat"],
            mock_params["lon"],
            mock_params["appid"],
        )
        == "Latitude input should be a float or an integer"
    )

    assert (
        airpyllution.get_air_pollution(
            mock_params["lat"],
            mock_incorrect_params["lon"],
            mock_params["appid"],
        )
        == "Longitude input should be a float or an integer"
    )

    assert (
        airpyllution.get_air_pollution(
            mock_params["lat"], mock_params["lon"], 0
        )
        == "API Key should be a string"
    )

    # Invalid input values
    assert (
        airpyllution.get_air_pollution(
            mock_incorrect_params["lat_oor"],
            mock_params["lon"],
            mock_params["appid"],
        )
        == "Enter valid latitude values (Range should be -90<Latitude<90)"
    )

    assert (
        airpyllution.get_air_pollution(
            mock_params["lat"],
            mock_incorrect_params["lon_oor"],
            mock_params["appid"],
        )
        == "Enter valid longitude values (Range should be -180<Longitude<180)"
    )

    # API error
    assert (
        airpyllution.get_air_pollution(
            mock_params["lat"],
            mock_params["lon"],
            mock_error_params["appid"],
        )
        == "An error occurred requesting data from the API"
    )

    assert (
        airpyllution.get_air_pollution(
            mock_params["lat"],
            mock_params["lon"],
            mock_incorrect_params["appid"],
        )
        == mock_invalid_message
    )

    assert (
        airpyllution.get_air_pollution(
            mock_params["lat"], mock_params["lon"], mock_params["appid"], 0
        )
        == "Figure title should be a string"
    )

    # Functionality check with mocked data
    fig = airpyllution.get_air_pollution(
        mock_params["lat"], mock_params["lon"], mock_params["appid"], "test"
    )
    # Check that the plot is a geographic scatter plot
    assert fig["data"][0]["type"] == "scattergeo"
    # Check that there are 8 pollutants
    assert len(fig["data"]) == 8
    # Check that the plot title matches fig_title
    assert fig["layout"]["title"]["text"] == "test"
    # Check that the legend title is correct
    assert fig["layout"]["legend"]["title"]["text"] == "Pollutant"
    # Check that the latitude value is correct on the plot
    assert floor(fig["data"][0]["lat"][0]) == floor(mock_params["lat"])
    # Check that the longitude value is correct on the plot
    assert floor(fig["data"][0]["lon"][0]) == floor(mock_params["lon"])


@patch("requests.get", side_effect=mocked_requests_get_pollution)
def test_pollution_forecast(mock_api_call):
    """Tests pollution forecast function"""

    # Invalid input type
    assert (
        airpyllution.get_pollution_forecast(
            mock_incorrect_params["lat"],
            mock_params["lon"],
            mock_params["appid"],
        )
        == "Latitude input should be a float"
    )

    assert (
        airpyllution.get_pollution_forecast(
            mock_params["lat"],
            mock_incorrect_params["lon"],
            mock_params["appid"],
        )
        == "Longitude input should be a float"
    )

    assert (
        airpyllution.get_pollution_forecast(
            mock_params["lat"], mock_params["lon"], 0
        )
        == "API Key should be a string"
    )

    # Invalid input values
    assert (
        airpyllution.get_pollution_forecast(
            mock_incorrect_params["lat_oor"],
            mock_params["lon"],
            mock_params["appid"],
        )
        == "Enter valid latitude values (Range should be -90<Latitude<90)"
    )

    assert (
        airpyllution.get_pollution_forecast(
            mock_params["lat"],
            mock_incorrect_params["lon_oor"],
            mock_params["appid"],
        )
        == "Enter valid longitude values (Range should be -180<Longitude<180)"
    )

    assert (
        airpyllution.get_pollution_forecast(
            mock_params["lat"],
            mock_params["lon"],
            mock_error_params["appid"],
        )
        == "An error occurred requesting data from the API"
    )

    assert (
        airpyllution.get_pollution_forecast(
            mock_params["lat"],
            mock_params["lon"],
            mock_incorrect_params["appid"],
        )
        == mock_invalid_message
    )

    # Functionality check with mocked data
    forecast_chart = airpyllution.get_pollution_forecast(
        mock_params["lat"], mock_params["lon"], mock_params["appid"]
    )

    assert forecast_chart.columns == 3
    assert len(forecast_chart.data) > 2
    assert len(forecast_chart.data) == 16
    assert (
        forecast_chart.title == "Pollutant concentration for the next 5 days"
    )
    chart_dict = forecast_chart.to_dict()
    assert chart_dict["facet"] == alt.Facet("Pollutants:N").to_dict()


src/airpyllution/utils.py
from datetime import datetime
import pandas as pd


def convert_unix_to_date(utime):
    """Returns formatted date time string

    Given a UNIX timestamp, this function reformats the timestamp
    to a string

    Parameters
    ----------
    utime : int
        UNIX timestamp, e.g. 1606488670
    Returns
    -------
    datetime
    Examples
    --------
    >>> date_conversion(1606488670)
    2020-11-27 17:00:00
    """
    ts = int(utime)
    return datetime.utcfromtimestamp(ts).strftime("%Y-%m-%d %H:%M:%S")


def convert_data_to_pandas(raw_data):
    """Converts API data from OpenWeatherMap to a pandas dataframe

    Parses the JSON data object and transforms it to a dataframe
    with a formatted date column

    Parameters
    ----------
    raw_data : JSON object

    Returns
    -------
    Pandas.dataframe
    Examples
    --------
    >>> convert_data_to_pandas(data)
    """
    if len(raw_data["list"]) > 1:
        data = pd.DataFrame.from_records(
            list(map(lambda x: x["components"], raw_data["list"]))
        )
        data["dt"] = list(
            map(lambda x: convert_unix_to_date(x["dt"]), raw_data["list"])
        )
    else:
        data = raw_data["coord"]
        data.update(raw_data["list"][0]["components"])
        data = pd.DataFrame.from_records([data])

    return data


src/AirPyllution/AirPyllution.py

def get_pollution_history(start_date, end_date, lat, lon, api_key):
    """Returns a dataframe of pollution history for a location between a specified date range
    
    Given a specified date range, the function makes an API request to the OpenWeather Air Pollution API and fetches
    historic pollution data for a given location.

    The function transforms the returned JSON object from the request into a Pandas dataframe.

    Note: Historical data is accessible from 27th November 2020

    Parameters
    ----------
    start_date : int
        start date of the time frame as a UNIX timestamp, e.g. 1606488670
    end_date : int
        end date of the time frame as a UNIX timetamp, e.g. 1606747870
    lat : float
        geographical latitude coordinate for the location
    lon : float
        geographical longitude coordinate for the location
    api_key: string
        OpenWeather API key
    Returns
    -------
    pandas.DataFrame
        a dataframe of the data returned from Air Pollution API - columns are as followed:
        ==========  ==============================================================
        date        int
        co          float: Carbon monoxide
        no          float: Nitrogen monoxide
        no2         float: Nitrogen dioxide
        o3          float: Ozone
        so2         float: Sulphur Dioxide
        pm2_5       float: Particulates 2.5
        pm10        float: Particulates 10
        nh3         float: Ammonia
        ==========  ==============================================================
    Examples
    --------
    >>> get_pollution_history(1606488670, 1606747870, 49.28, 123.12, "APIKEY_example")
    0 1606482000 270.367 5.867 43.184 4.783 14.544 13.448 15.524 0.289
    1 1606478400 280.38 8.605 42.155 2.459 14.901 15.103 17.249 0.162
    2 1606474800 293.732 13.523 41.47 1.173 15.14 17.727 19.929 0.072
    """ 

src/airpyllution/airpyllution.py
import requests
from airpyllution.utils import convert_data_to_pandas
import plotly.express as px
import altair as alt
import pandas as pd

alt.renderers.enable("mimetype")

# import constants
OPEN_WEATHER_MAP_URL = "http://api.openweathermap.org/data/2.5/air_pollution"


def get_pollution_history(start_date, end_date, lat, lon, api_key):
    """Returns a dataframe of pollution history for a location between
    a specified date range

    Given a specified date range, the function makes an API request to
    the OpenWeather Air Pollution API and fetches historic pollution data
    for a given location.

    The function transforms the returned JSON object from the request into
    a Pandas dataframe.

    Note: Historical data is accessible from 27th November 2020

    Parameters
    ----------
    start_date : int
        start date of the time frame as a UNIX timestamp, e.g. 1606488670
    end_date : int
        end date of the time frame as a UNIX timetamp, e.g. 1606747870
    lat : float
        geographical latitude coordinate for the location
    lon : float
        geographical longitude coordinate for the location
    api_key: string
        OpenWeather API key
    Returns
    -------
    pandas.DataFrame
        a dataframe of the data returned from the API.
        Columns are as followed:
        ==========  =================================================
        date        int
        co          float: Carbon monoxide
        no          float: Nitrogen monoxide
        no2         float: Nitrogen dioxide
        o3          float: Ozone
        so2         float: Sulphur Dioxide
        pm2_5       float: Particulates 2.5
        pm10        float: Particulates 10
        nh3         float: Ammonia
        ==========  ==================================================
    Examples
    --------
    >>> get_pollution_history(1606488670, 1606747870, 49.28, 123.12,
    "APIKEY_example")
    0 1606482000 270.367 5.867 43.184 4.783 14.544 13.448 15.524 0.289
    1 1606478400 280.38 8.605 42.155 2.459 14.901 15.103 17.249 0.162
    2 1606474800 293.732 13.523 41.47 1.173 15.14 17.727 19.929 0.072
    """

    # api_key = app.config["OPEN_WEATHER_MAP_API_KEY"]

    if not isinstance(lat, (float, int)):
        return "Latitude input should be a float"

    if not isinstance(lon, (float, int)):
        return "Longitude input should be a float"

    if not isinstance(start_date, int):
        return "start_date input should be an int"

    if not isinstance(end_date, int):
        return "end_date input should be an int"

    url = OPEN_WEATHER_MAP_URL + "/history"
    params = {
        "lat": lat,
        "lon": lon,
        "start": start_date,
        "end": end_date,
        "appid": api_key,
    }

    try:
        response = requests.get(url=url, params=params)
        response_obj = response.json()

        try:
            data = convert_data_to_pandas(response_obj)
            return data
        except Exception:
            if "cod" in response_obj:
                return response_obj["message"]
    except Exception:
        return "An error occurred requesting data from the API"


def get_air_pollution(lat, lon, api_key, fig_title=""):
    """Returns a map depicting varying pollution levels for a specified location.

    The function makes an API request to the OpenWeather Air Pollution
    API and fetches pollution data for a given location.

    The function transforms the returned JSON object from the request
    into a plotly plot.

    Parameters
    ----------
    lat : float
        geographical latitude coordinate for the location
    lon : float
        geographical longitude coordinate for the location
    api_key: string
        OpenWeather API key
    fig_title: string
        title of returned figure
    Returns
    -------
    plotly.graph_objs._figure.Figure

    Examples
    --------
    >>> get_air_pollution(49.2497, -123.1193, "APIKEY_example")
    """
    if not isinstance(lat, (float, int)):
        return "Latitude input should be a float or an integer"

    if not isinstance(lon, (float, int)):
        return "Longitude input should be a float or an integer"

    if not isinstance(api_key, str):
        return "API Key should be a string"

    if lat < -90.0 or lat > 90.0:
        return "Enter valid latitude values (Range should be -90<Latitude<90)"

    if lon < -180.0 or lon > 180.0:
        return (
            "Enter valid longitude values (Range should be -180<Longitude<180)"
        )

    if not isinstance(fig_title, str):
        return "Figure title should be a string"

    url = OPEN_WEATHER_MAP_URL
    params = {
        "lat": lat,
        "lon": lon,
        "appid": api_key,
    }

    try:
        response = requests.get(url=url, params=params)
        response_obj = response.json()
        try:
            data = convert_data_to_pandas(response_obj)
        except Exception:
            if "cod" in response_obj:
                return response_obj["message"]
    except Exception:
        return "An error occurred requesting data from the API"

    data = data.melt(
        id_vars=["lon", "lat"],
        value_vars=["co", "no", "no2", "o3", "so2", "pm2_5", "pm10", "nh3"],
    )

    data = data.replace(
        {
            "co": "CO",
            "no": "NO",
            "no2": "NO2",
            "o3": "O3",
            "so2": "SO2",
            "pm2_5": "PM2.5",
            "pm10": "PM10",
            "nh3": "NH3",
        }
    )

    fig = px.scatter_geo(
        data,
        lon="lon",
        lat="lat",
        color="variable",
        size="value",
        projection="natural earth",
        labels={
            "lon": "Longitude",
            "lat": "Latitude",
            "variable": "Pollutant",
            "value": "Conc. (µg/m^3)",
        },
        title=fig_title,
    )

    fig.update_layout(
        legend={"yanchor": "middle", "y": 0.72, "xanchor": "right", "x": 1.2},
        title={"y": 0.85, "x": 0.47, "xanchor": "center", "yanchor": "top"},
    )

    return fig


def get_pollution_forecast(lat, lon, api_key):
    """Returns a time series plot showing predicted pollutant levels
    for the next 5 days.

    Performs an API request to OpenWeather Air Pollution API,
    retrieves weather forecast for the next 5 days, and creates a time series
    graph of the pollutants with their concentration levels.

    Parameters
    ----------
    lat : float
        geographical latitude coordinate for the location
    lon : float
        geographical longitude coordinate for the location
    api_key: string
        OpenWeather API key

    Returns
    -------
    altair.Chart
        altair chart object with the x axis as time/UNIX timestamp and
        y axis as pollutant concentration.

    Examples
    --------
    >>> get_pollution_forecast(50, 50, "APIKEY_example")
    """
    if not isinstance(lat, (float, int)):
        return "Latitude input should be a float"

    if not isinstance(lon, (float, int)):
        return "Longitude input should be a float"

    if not isinstance(api_key, str):
        return "API Key should be a string"

    if lat < -90.0 or lat > 90.0:
        return "Enter valid latitude values (Range should be -90<Latitude<90)"

    if lon < -180.0 or lon > 180.0:
        return (
            "Enter valid longitude values (Range should be -180<Longitude<180)"
        )

    url = OPEN_WEATHER_MAP_URL + "/forecast"
    params = {"lat": lat, "lon": lon, "appid": api_key}

    try:
        response = requests.get(url=url, params=params)
        response_obj = response.json()
        try:
            data = convert_data_to_pandas(response_obj)
        except Exception:
            if "cod" in response_obj:
                return response_obj["message"]
    except Exception:
        return "An error occurred requesting data from the API"

    if len(data) >= 1:
        try:
            data = data.melt(
                id_vars=["dt"],
                value_vars=[
                    "co",
                    "no",
                    "no2",
                    "o3",
                    "so2",
                    "pm2_5",
                    "pm10",
                    "nh3",
                ],
                var_name="Pollutants",
                value_name="Concentration",
            )
            data["dt"] = pd.to_datetime(data["dt"])
            chart = (
                alt.Chart(data)
                .mark_line()
                .encode(
                    x=alt.X("dt", title="date"),
                    y=alt.Y("Concentration", title="Concentration"),
                    color=alt.Color("Pollutants"),
                )
                .properties(width=100, height=100)
                .facet(facet="Pollutants:N", columns=3)
                .resolve_axis(x="independent", y="independent")
                .resolve_scale(x="independent", y="independent")
                .properties(
                    title="Pollutant concentration for the next 5 days",
                )
                .configure_title(fontSize=24, anchor="middle")
            )
        except Exception:
            return "An error occured in plotting"
        return chart
    else:
        return "Insufficient data to forecast/plot."
